mail_notify_bounce/models/mail_thread.py

```diff
@@ -0,0 +1,66 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Lorenzo Battistini - Agile Business Group
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
+
+import logging
+from odoo import models, api, tools
+
+_logger = logging.getLogger(__name__)
+
+
+class MailThread(models.AbstractModel):
+    _inherit = 'mail.thread'
+
+    @api.model
+    def message_route(
+        self, message, message_dict, model=None, thread_id=None,
+        custom_values=None
+    ):
+        if self.env.context.get('fetchmail_server_id'):
+            fetchmail = self.env['fetchmail.server'].browse(
+                self.env.context['fetchmail_server_id'])
+            if fetchmail.bounce_notify_partner_ids:
+                message_id = message.get('Message-Id')
+                email_from = tools.decode_message_header(message, 'From')
+
+                references = tools.decode_message_header(message, 'References')
+                in_reply_to = tools.decode_message_header(
+                    message, 'In-Reply-To').strip()
+                thread_references = references or in_reply_to
+                msg_references = [
+                    ref for ref in
+                    tools.mail_header_msgid_re.findall(thread_references) if
+                    'reply_to' not in ref
+                ]
+                MailMessage = self.env['mail.message']
+                mail_messages = MailMessage.sudo().search([
+                    ('message_id', 'in', msg_references)], limit=1)
+                recipients = mail_messages.mapped('author_id')
+                recipients |= fetchmail.bounce_notify_partner_ids
+
+                email_from_localpart = (
+                    (
+                        tools.email_split(email_from) or ['']
+                    )[0].split('@', 1)[0].lower()
+                )
+                if (
+                    message.get_content_type() == 'multipart/report' or
+                    email_from_localpart == 'mailer-daemon'
+                ):
```

Comment:
detecting bounces is hard. I'd suggest to make this configurable somehow, maybe config parameters with regexes

Reply:
@hbrunn this is the same criterion used by odoo: https://github.com/odoo/odoo/blob/0783318f9745c4e213016b825e9bbd8e1e9bd546/addons/mail/models/mail_thread.py#L1053
Comment added